macros/graph/PGplot.pl

```diff
@@ -0,0 +1,1914 @@
+		$point += $delta;
+	} until (-$point < $min && $point > $max);
+	return \@ticks;
+}
+
+sub grid {
+	my $self = shift;
+	return $self->get('xmajor', 'xminor', 'xticks', 'ymajor', 'yminor', 'yticks');
+}
+
+sub bounds {
+	my $self = shift;
+	return $self->{xaxis}{min}, $self->{yaxis}{min}, $self->{xaxis}{max}, $self->{yaxis}{max};
+}
+
+=head1 DATA OBJECT
+
+This object holds data about the different types of elements that can be added
+to a PGplot graph. This is a hash with some helper methods. Data objects are created
+and modified using the PGplot methods, and do not need to generally be
+modified in a PG problem.  Each PG add method returns the related data object which
+can be used if needed.
+
+Each data object contains the following:
+
+=over 5
+
+=item name
+
+The name is used to identify what type of data is being stored,
+such as a function, dataset, label, etc.
+
+=item x
+
+The array of the data points x-value.
+
+=item y
+
+The array of the data points y-value.
+
+=item function
+
+A function (stored as a hash) to generate the x and y data points.
+
+=item styles
+
+An hash of different style options and values that can be used
+to store additional data for things like color, width, etc.
+
+=back
+
+=head1 USAGE
+
+The main methods for adding data and accessing the data are:
+
+=over 5
+
+=item C<$data-E<gt>name>
+
+Sets, C<$data-E<gt>name($string)>, or gets C<$data-E<gt>name> the name of the data object.
+
+=item C<$data-E<gt>add>
+
+Adds a single data point, C<$data-E<gt>add($x, $y)>, or adds multiple data points,
+C<$data-E<gt>add([$x1, $y1], [$x2, $y2], ..., [$xn, $yn])>.
+
+=item C<$data-E<gt>set_function>
+
+Configures a function to generate data points. C<sub_x> and C<sub_y> are are perl subroutines.
+
+    $data->set_function(
+        sub_x => sub { return $_[0]; },
+        sub_y => sub { return $_[0]**2; },
+        min   => -5,
+        max   => 5,
+    );
+
+The number of steps used to generate the data is a style and needs to be set separately.
+
+    $data->style(steps => 50);
+
+=item C<$data-E<gt>gen_data>
+
+Generate the data points from a function. This can only be done when there is no data, so
+once the data has been generated this will do nothing (to avoid generating data again).
+
+=item C<$data-E<gt>size>
+
+Returns the current number of points being stored.
+
+=item C<$data-E<gt>x> and C<$data-E<gt>y>
+
+Without any inputs, these return either the x array or y array of data points being stored.
+A single input can be used to return only the n-th data point, C<$data-E<gt>x($n)>.
+
+=item C<$data-E<gt>style>
+
+Sets or gets style information. Use C<$data-E<gt>style($name)> to get the style value of a single
+style name. C<$data-E<gt>style> will returns a reference to the full style hash. Last, input a hash
+to add / change the styles.
+
+    $data->style(color => 'blue', width => 3);
+
+=back
+
+=cut
+
+package PGplot::Data;
+
+sub new {
+	my $class = shift;
```

Comment:
Unpack the arguments here by adding `%options`, and bless `$self` below directly.